server/channels/store/sqlstore/retention_policy_store.go

```diff
@@ -958,31 +1046,127 @@ func genericRetentionPoliciesDeletion(
+			query = `
+			DELETE FROM ` + r.Table + ` WHERE ` + primaryKeysStr + ` IN (
+			` + query + `
+			) RETURNING ` + r.Table + `.` + r.PrimaryKeys[0]
+			var rows *sql.Rows
+			rows, err = txn.Query(query, args...)
+			if err != nil {
+				return 0, errors.Wrap(err, "failed to delete "+r.Table)
+			}
+
+			defer rows.Close()
+			ids := []string{}
+			for rows.Next() {
+				var id string
+				if err = rows.Scan(&id); err != nil {
+					return 0, errors.Wrap(err, "unable to scan from rows")
+				}
+				ids = append(ids, id)
+			}
+			if err = rows.Err(); err != nil {
+				return 0, errors.Wrap(err, "failed while iterating over rows")
+			}
+			rowsAffected = int64(len(ids))
+
+			if len(ids) > 0 {
+				retentionIdsRow := model.RetentionIdsForDeletion{
+					TableName: r.Table,
+					Ids:       ids,
+				}
+				err = insertRetentionIdsForDeletion(txn, &retentionIdsRow, s)
+				if err != nil {
+					return 0, err
+				}
+			}
+		} else {
+			retentionIdsRow := model.RetentionIdsForDeletion{
+				TableName: r.Table,
+				Ids:       []string{},
+			}
+			// 1. Select rows that will be deleted
+			if err = txn.Select(&retentionIdsRow.Ids, query, args...); err != nil {
+				return 0, err
+			}
+
+			if len(retentionIdsRow.Ids) > 0 {
+				err = insertRetentionIdsForDeletion(txn, &retentionIdsRow, s)
+				if err != nil {
+					return 0, err
+				}
+
+				// MySQL does not support the LIMIT clause in a subquery with IN
+				clauses := make([]string, len(r.PrimaryKeys))
+				for i, key := range r.PrimaryKeys {
+					clauses[i] = r.Table + "." + key + " = A." + key
+				}
+				joinClause := strings.Join(clauses, " AND ")
+				query = `
+				DELETE ` + r.Table + ` FROM ` + r.Table + ` INNER JOIN (
+				` + query + `
+				) AS A ON ` + joinClause
+
+				// 3. Delete from Parent table
+				var result sql.Result
+				result, err = txn.Exec(query, args...)
+				if err != nil {
+					return 0, errors.Wrap(err, "failed to delete "+r.Table)
+				}
+
+				rowsAffected, err = result.RowsAffected()
+				if err != nil {
+					return 0, errors.Wrap(err, "failed to get rows affected for "+r.Table)
+				}
+			}
+		}
+		if err = txn.Commit(); err != nil {
+			return 0, err
+		}
 	} else {
-		// MySQL does not support the LIMIT clause in a subquery with IN
-		clauses := make([]string, len(r.PrimaryKeys))
-		for i, key := range r.PrimaryKeys {
-			clauses[i] = r.Table + "." + key + " = A." + key
+		if s.DriverName() == model.DatabaseDriverPostgres {
+			primaryKeysStr := "(" + strings.Join(r.PrimaryKeys, ",") + ")"
+			query = `
+			DELETE FROM ` + r.Table + ` WHERE ` + primaryKeysStr + ` IN (
+			` + query + `
+			)`
+		} else {
+			// MySQL does not support the LIMIT clause in a subquery with IN
+			clauses := make([]string, len(r.PrimaryKeys))
+			for i, key := range r.PrimaryKeys {
+				clauses[i] = r.Table + "." + key + " = A." + key
+			}
+			joinClause := strings.Join(clauses, " AND ")
+			query = `
+			DELETE ` + r.Table + ` FROM ` + r.Table + ` INNER JOIN (
```

Comment:
Can you mention in a comment what the final query will look like? I am having a bit of trouble imagining.

Reply:
This is MySQL:
Channel Retention Policy
```
DELETE Posts FROM Posts INNER JOIN (
    SELECT Posts.Id 
    FROM Posts 
    INNER JOIN Channels ON Posts.ChannelId = Channels.Id 
    INNER JOIN RetentionPoliciesChannels ON Posts.ChannelId = RetentionPoliciesChannels.ChannelId 
    INNER JOIN RetentionPolicies ON RetentionPoliciesChannels.PolicyId = RetentionPolicies.Id 
    WHERE (RetentionPolicies.PostDuration >= ? AND 1692210977000 - Posts.CreateAt > RetentionPolicies.PostDuration * 86400000) 
    LIMIT 100
) AS A ON Posts.Id = A.Id;
args:[0]
```

Team retention policy
```
DELETE Posts FROM Posts INNER JOIN (
    SELECT Posts.Id 
    FROM Posts 
    INNER JOIN Channels ON Posts.ChannelId = Channels.Id 
    LEFT JOIN RetentionPoliciesChannels ON Posts.ChannelId = RetentionPoliciesChannels.ChannelId 
    INNER JOIN RetentionPoliciesTeams ON Channels.TeamId = RetentionPoliciesTeams.TeamId 
    INNER JOIN RetentionPolicies ON RetentionPoliciesTeams.PolicyId = RetentionPolicies.Id 
    WHERE (RetentionPoliciesChannels.PolicyId IS NULL AND RetentionPoliciesTeams.PolicyId = RetentionPolicies.Id) 
    AND (RetentionPolicies.PostDuration >= ? AND 1692210977000 - Posts.CreateAt > RetentionPolicies.PostDuration * 86400000) 
    LIMIT 100
) AS A ON Posts.Id = A.Id; 
args:[0]
```

Global retention policy
```
DELETE Posts FROM Posts INNER JOIN (
    SELECT Posts.Id 
    FROM Posts 
    INNER JOIN Channels ON Posts.ChannelId = Channels.Id 
    LEFT JOIN RetentionPoliciesChannels ON Posts.ChannelId = RetentionPoliciesChannels.ChannelId 
    LEFT JOIN RetentionPoliciesTeams ON Channels.TeamId = RetentionPoliciesTeams.TeamId 
    LEFT JOIN RetentionPolicies ON RetentionPoliciesChannels.PolicyId = RetentionPolicies.Id 
    WHERE (RetentionPoliciesChannels.PolicyId IS NULL AND RetentionPoliciesTeams.PolicyId IS NULL) 
    AND Posts.CreateAt < ? 
    LIMIT 100
) AS A ON Posts.Id = A.Id;
args: [1555473600000]
```